### ISSUE TITLE
exclude_times functionality

### ISSUE THREAD

**Issue by djrobby** [haydenrou/slotty#2]
First off, thank you so much for such a handy utility!

I'm running into a bit of an issue with the following code not returning any available time slots. From your example, it looks like the `exclude_times` attribute accepts an `Array` of `Hashes`. Can you please advise if I'm overlooking anything? Thanks in advance!

```ruby
resource_busy_times = [
  [10, 00, 10, 45],
  [11, 30, 12, 15],
  [12, 15, 13, 00],
  [13, 00, 13, 45],
  [14, 00, 14, 45],
  [15, 00, 17, 00],
  [17, 00, 17, 30],
  [19, 00, 19, 45]
]
busy_times = []
resource_busy_times.each do |t|
  busy_times << { start: Time.new(2020, 05, 22, t[0], t[1]), end: Time.new(2020, 05, 22, t[2], t[3]) }
end

# irb> puts busy_times
# [{:start=>2020-05-22 10:00:00 -0400, :end=>2020-05-22 10:45:00 -0400},
#  {:start=>2020-05-22 11:30:00 -0400, :end=>2020-05-22 12:15:00 -0400},
#  {:start=>2020-05-22 12:15:00 -0400, :end=>2020-05-22 13:00:00 -0400},
#  {:start=>2020-05-22 13:00:00 -0400, :end=>2020-05-22 13:45:00 -0400},
#  {:start=>2020-05-22 14:00:00 -0400, :end=>2020-05-22 14:45:00 -0400},
#  {:start=>2020-05-22 15:00:00 -0400, :end=>2020-05-22 17:00:00 -0400},
#  {:start=>2020-05-22 17:00:00 -0400, :end=>2020-05-22 17:30:00 -0400},
#  {:start=>2020-05-22 19:00:00 -0400, :end=>2020-05-22 19:45:00 -0400}]

Slotty.get_slots(
  from: Time.new(2020, 05, 22, 10, 00),
  to: Time.new(2020, 05, 22, 17, 00),
  slot_length: 60*30,
  timeframe: 60*15,
  exclude_times: busy_times
)

# irb> []
```
I was hoping it would output the following `30 minute` slots:
```ruby
# [{:start=>2020-05-22 10:45:00 -0400, :end=>2020-05-22 11:15:00 -0400},
#  {:start=>2020-05-22 11:00:00 -0400, :end=>2020-05-22 11:30:00 -0400}]
```

**Comment by djrobby** [comment]
Nevermind, all set! I ended up modifying `lib/slotty.rb` for my use case as below:

**From**
```ruby
# lib/slotty.rb
potential_slot = from..(from + slot_length)
```
**To**
```ruby
# lib/slotty.rb
potential_slot = from+1..(from + slot_length-1)
```
Thank you again! Closing issue.

**Comment by haydenrou** [comment]
Hi @djrobby , glad you're finding use of this gem. I built this with the intention of adding more functionality and tying up any loose ends when I get to using it with a current project, likely in the next month or so.

I'm glad you've brought this up, I believe I need to implement a workaround for when two periods touch each other i.e. 10:30 - 11:00 and 11:00 - 11:30. I think this currently doesn't work as intended.

I'm going to reopen this issue and look into it soon

**Comment by djrobby** [comment]
Hi @haydenrou, thank you for getting back to me! The more I reviewed your algorithm, the more I was convinced that it needed to stay the way it was. So, I reverted my changes to `lib/slotty.rb` and instead, did the following:

```ruby
bt = [
  ["10:00:00", "10:30:00"],
  ["11:00:00", "11:30:00"],
  ["12:00:00", "12:30:00"],
  ["13:00:00", "13:30:00"],
  ["14:00:00", "14:30:00"],
  ["15:00:00", "15:30:00"],
  ["17:30:00", "18:00:00"],
  ["18:00:00", "19:00:00"]
]
busy_times = []
bt.each do |t|
  busy_times << { start: Time.parse([date, t[0]].join(' '))+1, end: Time.parse([date, t[1]].join(' '))-1 }
end
```
I simply offset the `busy_times` with `+1 second` to start and `-1 to end`. Your algorithm is pristine! Thank you again.

**Comment by haydenrou** [comment]
I'm glad you have a work-around, but in my head I still think this should be dealt with by the gem. I'll update & close this issue when I submit a fix.

Cheers

**Comment by haydenrou** [comment]
@djrobby I have opened a PR which will fix your above issue, if you're still using the gem!

You'll have to change the named attributes passed in, but it's all documented in the README and spec-tests in pr #3 .

I've got a couple more spec tests to write which I'll do tomorrow and then push up and version the gem.

Thanks again for commenting, it's always nice to see someone using something you've made!
